### PR TITLE
pilot-core: automatic modes introduction

### PR DESCRIPTION
**Pull request by niewim19**
## Description

Preliminary automatic modes introduction:
 - more modularized control module for easy RC override possibility
 - working implementation of `takeoff`, `hover` and `landing` procedure
 - changes of z accel bias estimation

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (zturn drone).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.
- [ ] - [ ]

### CODE REVIEW

File: quadcontrol/control.c [hunk 1]
```diff
@@ -57,6 +57,9 @@
 /* rcbus trigger thresholds for manual switches SWA/SWB/SWC/SWD */
 #define RCTHRESH_HIGH ((95 * (MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE)) / 100 + MIN_CHANNEL_VALUE) /* high position threshold */
 #define RCTHRESH_LOW  ((5 * (MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE)) / 100 + MIN_CHANNEL_VALUE)  /* low position threshold */
```
Comment (HBuczynski) [inline quadcontrol/control.c:59]: I know that it is not a part of this PR, but you can add info that it is strictly related to channels
`RC_CHANNEL_THR_HIGH` & `RC_CHANNEL_THR_LOW`

File: quadcontrol/control.c [hunk 2]
```diff
@@ -171,53 +181,155 @@ static int quad_motorsCtrl(float throttle, int32_t alt, const quad_att_t *att, c
 }
 
 
+/* Overrides parameters selected with `ovrdFlags` with RC values */
+static void quad_rcOverride(quad_att_t *att, float *throttle, uint32_t ovrdFlags)
```
Comment (HBuczynski) [inline quadcontrol/control.c:185]:
```suggestion
static void quad_rcOverride(quad_att_t *att, float *throttle, uint32_t flags)
```

File: quadcontrol/control.c [hunk 3]
```diff
@@ -171,53 +181,155 @@ static int quad_motorsCtrl(float throttle, int32_t alt, const quad_att_t *att, c
...
+	/* +-0.5 radian ~= +- 28 degrees */
+	if ((ovrdFlags & RC_OVRD_LEVEL) != 0 && att != NULL) {
+		att->roll = (rcRoll - ((MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE) / 2)) / 1000.f;
+		/* minus in pitch is to compensate for rc transmitter sign: stick up => higher channel value => negative pitch change */
+		att->pitch = -(rcPitch - ((MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE) / 2)) / 1000.f;
+	}
+
+	/* delta yaw in +-500 milliradians range */
+	if ((ovrdFlags & RC_OVRD_YAW) != 0 && att != NULL) {
+		att->yaw += (float)(rcYaw - ((MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE) / 2)) / 1000;
```
Comment (HBuczynski) [inline quadcontrol/control.c:205]:
```suggestion
		att->yaw += (float)(rcYaw - ((MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE) / 2)) / 1000.f;
```

File: quadcontrol/control.c [hunk 4]
```diff
@@ -278,96 +319,94 @@ static int quad_landing(const flight_mode_t *mode)
...
-		att.pitch = -(rcPitch - ((MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE) / 2)) / 1000.f; /* +-0.5 radian ~= +- 28 degrees */
+		/* SWC == LOW (or illegal value) -> stabilize */
+		if (stickSWC < (0.1 * (MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE)) || stickSWC > MAX_CHANNEL_VALUE) {
+			att.yaw = measure.yaw;
+			alt = setAlt = measure.enuZ * 1000;
 
-		/* Update target yaw/altitude only if we are not in althold */
-		if (althold == false) {
-			throttle = hoverThrtlEst * ((float)rcThrottle / (MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE) + 0.5); /* throttle now in range (0.5, 1.5) of hover throttle */
+			quad_common.pids[pwm_alt].flags |= PID_IGNORE_P | PID_IGNORE_I | PID_IGNORE_D;
```
Comment (HBuczynski) [inline quadcontrol/control.c:473]: Please add a comment why you want to ignore PID

File: quadcontrol/control.c [hunk 5]
```diff
@@ -278,96 +319,94 @@ static int quad_landing(const flight_mode_t *mode)
...
+		else if (stickSWC > (0.9 * (MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE))) {
+			/* Do not use I altitude pid if there is too big difference between current alt and set alt */
+			if (fabs(measure.enuZ * 1000 - setAlt) > 1000) {
+				quad_common.pids[pwm_alt].flags |= PID_IGNORE_I;
 			}
+			else {
+				quad_common.pids[pwm_alt].flags &= ~PID_IGNORE_I;
+			}
+
+			setAlt = alt + 5000;
```
Comment (HBuczynski) [inline quadcontrol/control.c:497]: `5000` is a magic number

File: quadcontrol/control.c [hunk 6]
```diff
@@ -216,36 +216,105 @@ static void quad_rcOverride(quad_att_t *att, float *throttle, uint32_t ovrdFlags
 
 static int quad_takeoff(const flight_mode_t *mode)
 {
+	const float idleTime = mode->takeoff.idleTime;
+	const float spoolTime = mode->takeoff.spoolTime;
+	const float liftTime = mode->takeoff.liftTime;
+
+	const float hoverThrottle = quad_common.throttle.max; /* set hoverThrottle as throttle.max */
+	const int32_t startAlt = -5000;                       /* starting with negative altitude to lower the throttle */
```
Comment (HBuczynski) [inline quadcontrol/control.c:224]: You can make `5000` as a define.

File: quadcontrol/control.h [hunk 7]
```diff
@@ -27,8 +27,10 @@ typedef enum { /* Basic modes:  */ flight_idle = 0, flight_disarm, flight_arm,
 
 
 typedef struct {
-	int32_t alt; /* altitude in 1E-3 [m] (millimetres) above MSL */
-	float time;  /* time of engine spool up in milliseconds */
+	int32_t alt;   /* altitude in 1E-3 [m] (millimetres) above MSL */
+	int32_t idleTime;  /* relaxation time (milliseconds) after engines startup */
+	int32_t spoolTime; /* time (milliseconds) to reach hover throttle */
+	int32_t liftTime;  /* time (milliseconds) to reach PID controlled throttle values */
```
Comment (HBuczynski) [inline quadcontrol/control.h:33]: Please use `time_t` and update parsing functions.

File: quadcontrol/control.h [hunk 8]
```diff
@@ -53,7 +55,9 @@ typedef struct {
 
 
 typedef struct {
-	time_t time; /* time of engine spool down in milliseconds */
+	int32_t descent; /* Descent speed in millimeters per second */
+	int32_t diff;    /* Minimum altitude difference to occur when we suspect landing is done */
+	int32_t timeout; /* Time threshold of `diff` persistence to decide that landing is finished */
```
Comment (HBuczynski) [inline quadcontrol/control.h:60]: Please use `time_t`, update parsing functions and add units in the comment

**Review by HBuczynski — CHANGES_REQUESTED** [review]

**Comment by niewim19** [comment]
- added macros for magic numbers
- made time types consistent in using `time_t` type
- added necessary comments in conde

File: quadcontrol/control.c [hunk 9]
```diff
@@ -171,53 +185,155 @@ static int quad_motorsCtrl(float throttle, int32_t alt, const quad_att_t *att, c
...
+}
+
+
 /* Handling flight modes */
 
 static int quad_takeoff(const flight_mode_t *mode)
 {
+	const float idleTime = mode->takeoff.idleTime;
+	const float spoolTime = mode->takeoff.spoolTime;
+	const float liftTime = mode->takeoff.liftTime;
```
Comment (HBuczynski) [inline quadcontrol/control.c:225]: Why `float` instead of `time_t`?

**Review by HBuczynski — APPROVED** [review]
LGTM

**Pull request merged**